📖 Refactor InfraMachine contract (#11223)
kubernetes-sigs/cluster-api: docs/book/src/developer/core/controllers/machine.md

# Machine Controller

The Machine controller is responsible for reconciling the Machine resource.

In order to allow Machine provisioning on different type of infrastructure, The Machine resource references
an Machine object, e.g. AWSMachine, GCMachine etc.

> The [InfraMachine resource contract](../../providers/contracts/infra-machine.md) defines a set of rules a provider is expected to comply in order to allow

Author: fixed all around

the expected interactions with the Machine controller.

Among those rules:
- InfraMachine MUST report a [provider ID](../../providers/contracts/infra-machine.md#inframachine-provider-id) for the Machine

> - InfraMachine SHOULD define a [failure domain](../../providers/contracts/infra-machine.md#inframachine-failure-domain) where machines should be placed in

Reviewer: ```suggestion
- InfraMachine SHOULD define a [failure domain](../../providers/contracts/infra-machine.md#inframachine-failure-domain) field that decides where machines are placed in
```
Maybe. Sould define a failureDomain sounds a bit unclear

Otherwise it sounds like the InfraMachine comes up with the failureDomain, but I think it's the other way around

Author: Rephrased, PTAL

Author: If what I'm proposing is ok,  I'm going to open an issue to track this + brings this up in the office hours

Reviewer: In general sounds good to me.

I assume we still have use cases going forward where no failureDomain is specified in Machine, then the infra provider just picks one and we want to surface that back?

Author: https://github.com/kubernetes-sigs/cluster-api/issues/11232

- InfraMachine SHOULD surface machine's [addresses](../../providers/contracts/infra-machine.md#inframachine-addresses) to help operators when troubleshooting issues

> - InfraMachine MUST report when Cluster's infrastructure is [fully provisioned](../../providers/contracts/infra-machine.md#inframachine-initialization-completed)

Reviewer: ```suggestion
- InfraMachine MUST report when Machine's infrastructure is [fully provisioned](../../providers/contracts/infra-machine.md#inframachine-initialization-completed)
```

> - InfraMachine MUST report when Machine's's infrastructure is [fully provisioned](../../providers/contracts/infra-machine.md#inframachine-initialization-completed)

Reviewer: ```suggestion
- InfraMachine MUST report when Machine's infrastructure is [fully provisioned](../../providers/contracts/infra-machine.md#inframachine-initialization-completed)
```

- InfraMachine SHOULD report [conditions](../../providers/contracts/infra-machine.md#inframachine-conditions)
- InfraMachine SHOULD report [terminal failures](../../providers/contracts/infra-machine.md#inframachine-terminal-failures)

Similarly, in order to support different machine bootstrappers, The Machine resource references

> an BootstrapConfig object, e.g. KubeadmBoostrapConfig etc.

Reviewer: ```suggestion
a BootstrapConfig object, e.g. KubeadmBootstrapConfig etc.
```

> The [BootstrapConfig resource contract](../../providers/contracts/bootstrap-config.md) defines a set of rules a provider is expected to comply in order to allow

Reviewer: ```suggestion
The [BootstrapConfig resource contract](../../providers/contracts/bootstrap-config.md) defines a set of rules a provider is expected to comply with in order to allow
```

the expected interactions with the Machine controller.

Considering all the info above, the Machine controller's main responsibilities are:

* Setting an OwnerReference on the infrastructure object referenced in `Machine.spec.infrastructureRef`.
* Setting an OwnerReference on the bootstrap object referenced in `Machine.spec.bootstrap.configRef`.
* Keeping the Machine's status in sync with the InfraMachine and BootstrapConfig's status.
  * Finding Kubernetes nodes matching the expected providerID in the workload cluster.
  * Setting NodeRefs to be able to associate machines and Kubernetes nodes.
  * Monitor Kubernetes nodes and propagate labels to them.
* Cleanup of all owned objects so that nothing is dangling after deletion.

>   * Drain nodes and wait for volume being detached by the CSI provider.

Reviewer: ```suggestion
  * Drain nodes and wait for volumes being detached by CSI plugins.
```
I think they are not calling them providers
